common/src/main/java/org/apache/celeborn/common/network/client/TransportResponseHandler.java

```diff
@@ -39,22 +48,73 @@
 public class TransportResponseHandler extends MessageHandler<ResponseMessage> {
   private static final Logger logger = LoggerFactory.getLogger(TransportResponseHandler.class);
 
+  private final TransportConf conf;
   private final Channel channel;
 
   private final Map<StreamChunkSlice, ChunkReceivedCallback> outstandingFetches;
 
   private final Map<Long, RpcResponseCallback> outstandingRpcs;
-  private final Map<Long, RpcResponseCallback> outstandingPushes;
+  private final Map<Long, PushRequestInfo> outstandingPushes;
 
   /** Records the time (in system nanoseconds) that the last fetch or RPC request was sent. */
   private final AtomicLong timeOfLastRequestNs;
 
-  public TransportResponseHandler(Channel channel) {
+  private final ScheduledExecutorService pushTimeoutChecker;
+  private final long pushTimeoutCheckerInterval;
+
+  public TransportResponseHandler(TransportConf conf, Channel channel) {
+    this.conf = conf;
     this.channel = channel;
     this.outstandingFetches = new ConcurrentHashMap<>();
     this.outstandingRpcs = new ConcurrentHashMap<>();
     this.outstandingPushes = new ConcurrentHashMap<>();
     this.timeOfLastRequestNs = new AtomicLong(0);
+    pushTimeoutCheckerInterval = conf.pushDataTimeoutCheckIntervalMs();
+    pushTimeoutChecker =
+        ThreadUtils.newDaemonSingleThreadScheduledExecutor("push-timeout-checker-" + this);
+    pushTimeoutChecker.scheduleAtFixedRate(
+        new Runnable() {
+          @Override
+          public void run() {
+            failExpiredPushRequest();
+          }
+        },
+        pushTimeoutCheckerInterval,
+        pushTimeoutCheckerInterval,
+        TimeUnit.MILLISECONDS);
+  }
+
+  public void failExpiredPushRequest() {
+    long currentTime = System.currentTimeMillis();
+    for (Map.Entry<Long, PushRequestInfo> entry : outstandingPushes.entrySet()) {
+      try {
+        long requestId = entry.getKey();
+        PushRequestInfo info = entry.getValue();
+        if (info.pushDataTimeout > 0) {
+          if (info.pushTime != -1 && (currentTime - info.pushTime > info.pushDataTimeout)) {
+            if (info.callback != null) {
```

Comment:
There can be concurrent issues here. First we should use entry iterator instead of ```entry : outstandingPushes.entrySet```; Second we should remove from the outstandingPushes here; Third we check the return value of remove, if not null, do the other things.